src/Development/IDE/Core/Rules.hs

```diff
@@ -518,7 +502,12 @@ typeCheckRuleDefinition hsc pm generateArtifacts = do
   addUsageDependencies $ liftIO $ do
     res <- typecheckModule defer hsc pm
     case res of
-      (diags, Just (hsc,tcm)) | DoGenerateInterfaceFiles <- generateArtifacts -> do
+      (diags, Just (hsc,tcm))
+        | DoGenerateInterfaceFiles <- generateArtifacts
+        -- Don't save interface files for modules that compiled due to defering
+        -- type errors, as we won't get proper diagnostics if we load these from
+        -- disk
+        , not $ tmrDeferedError tcm -> do
```

Comment:
Shouldn't be to much of an issue since the result of `GetModIface` will still be cached in the shake graph?

Reply:
> Would it be a good idea to persist a list of warnings/errors next in or next to the interface file?

I would find this useful to integrate `.hie` files. Right now I can use the persisted `.hie` file to immediately give hover/definition/etc. responses, but I need to trigger a typecheck anyway in order to get up to date diagnostics.

Would the logic to check the freshness of the `.hi` file also suffice to check the freshness of persisted diagnostics?

Reply:
Will the result of `GetModIface` get cached if there is no interface file to store in the graph? 

The `.hie` freshness logic used to delegate to the `.hi` freshness logic, so it should suffice:

https://github.com/digital-asset/ghcide/blob/d4fd99edb39949b2cbc7e6d02a8418edc3335d8b/src/Development/IDE/Core/Rules.hs#L157-L164

Reply:
> Will the result of GetModIface get cached if there is no interface file to store in the graph?

The interface file is still generated and returned by `regenerateHiFile`, it will just not be persisted on disk, so won't be available for the next ghcide run.

> The .hie freshness logic used to delegate to the .hi freshness logic, so it should suffice:

I was talking wrt the freshness logic for the hypothetical persisted diagnostics.

Reply:
The name of the `generateAndWriteHiFile` function is a bit of a misnomer, since it doesn't really generate the interface, just writes it out after extracting it from the `HomeModInfo`.

https://github.com/digital-asset/ghcide/blob/master/src/Development/IDE/Core/Compile.hs#L301

Reply:
Done.